waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -948,16 +967,34 @@
   (get-services [this]
     (get-services this))
 
-  (kill-instance [this {:keys [id service-id] :as instance}]
+  (kill-instance [this {:keys [id service-id prepared-to-scale-down-at] :as instance}]
     (ss/try+
-      (let [service (service-id->service this service-id)]
-        (kill-service-instance this instance service)
-        (scheduler/log-service-instance instance :kill :info)
-        {:instance-id id
-         :killed? true
-         :message "Successfully killed instance"
-         :service-id service-id
-         :status http-200-ok })
+      (let [service (service-id->service this service-id)
+            desc (service-id->service-description-fn service-id)
+            bypass-enabled? (sd/service-description-bypass-enabled? desc)
+            first-phase? (and bypass-enabled? (nil? prepared-to-scale-down-at))
+            second-phase? (and bypass-enabled? (some? prepared-to-scale-down-at))
+            kill-result (cond first-phase?
+                              (do
+                                (mark-pod-for-scale-down this instance)
+                                {:marked-prepared-for-scale-down true
+                                 :message "Successfully annotated pod to be prepared for scale down"})
+                              second-phase?
```

Comment:
Don't we also need to check whether time has elapsed?

Reply:
This is done in the `find-killable-instance` in the `responder`. I did it this way because it didn't make sense to return an instance that was not killable based the timeout not getting reached in that fn.